### PR TITLE
C#: Properly dispose diagnostic writer objects

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

File: csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs [hunk 1]
```diff
@@ -370,6 +370,20 @@ void exitCallback(int ret, string msg, bool silent)
             }
         });
 
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
Comment (michaelnebel) [inline csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs:376]: Why is this needed?

File: csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs [hunk 2]
```diff
@@ -370,6 +370,20 @@ void exitCallback(int ret, string msg, bool silent)
...
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                (diagnostics as IDisposable)?.Dispose();
```
Comment (michaelnebel) [inline csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs:383]: Maybe let `IDiagnosticsWriter` implement `IDisposable` (the we should also get code quality warnings if the upcast `DiagnosticsStream` -> `IDiagnosticsWriter` hides a missing disposal). It will require adding a dummy dispose in one the test class, but that should be it.

**Comment by michaelnebel** [comment]
Also, very good catch. How did you discover this?

File: csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs (hunk 1 above)
Comment (tamasvajk) [reply csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs:376]: This is suggested by [CA1816](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1816). This is implementing the dispose pattern in case of non-sealed classes.

**Comment by tamasvajk** [comment]
> Also, very good catch. How did you discover this?

I was checking for https://github.com/github/codeql/pull/16021 how we write diagnostic messages, and I found that we have a disposable-like field. Also, I think it was only working before, because we call `.Flush()` explicitly in `DiagnosticsStream.AddEntry`.

File: csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs (hunk 2 above)
Comment (michaelnebel) [reply csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs:383]: Maybe we could argue that the naming of the class and interface could be changed a bit (scope creep).
I think we should rename `DiagnosticsStream` to `DiagnosticsWriter` and let `IDiagnosticsWriter` implement `IDisposable`.
My guess is that the only reason the interface was introduced was to allow stubbing and it seems very specific to the writer/serialize usecase.

File: csharp/extractor/Semmle.Util/ToolStatusPage/DiagnosticsStream.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Semmle.Util
+{
+    /// <summary>
+    /// A wrapper around an underlying <see cref="StreamWriter" /> which allows
+    /// <see cref="DiagnosticMessage" /> objects to be serialized to it.
+    /// </summary>
+    public sealed class DiagnosticsStream : IDiagnosticsWriter, IDisposable
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Util/ToolStatusPage/DiagnosticsStream.cs:12]:
```suggestion
    public sealed class DiagnosticsStream : IDiagnosticsWriter
```
